src/gtest/test_libsnark.cpp

```diff
@@ -0,0 +1,470 @@
+    ASSERT_TRUE(b3a == b3);
+    ASSERT_TRUE(!(b3a.is_zero()));
+
+    mpz_t m3;
+    mpz_init(m3);
+    b3.to_mpz(m3);
+    bigint<3> b3b { m3 };
+    ASSERT_TRUE(b3b == b3);
+
+    bigint<2> quotient;
+    bigint<2> remainder;
+    bigint<3>::div_qr(quotient, remainder, b3, b2);
+    ASSERT_TRUE(quotient.num_bits() < GMP_NUMB_BITS);
+    ASSERT_TRUE(quotient.as_ulong() == b1.as_ulong());
+    bigint<1> b1inc = bigint<1>("76749408");
+    bigint<1> b1a = quotient.shorten(b1inc, "test");
+    ASSERT_TRUE(b1a == b1);
+    ASSERT_TRUE(remainder.is_zero());
+    remainder.limit(b2, "test");
+
+    try {
+        (void)(quotient.shorten(b1, "test"));
+        ASSERT_TRUE(false);
+    } catch (std::domain_error) {}
+    try {
+        remainder.limit(remainder, "test");
+        ASSERT_TRUE(false);
+    } catch (std::domain_error) {}
+
+    bigint<1> br = bigint<1>("42");
+    b3 += br;
+    ASSERT_TRUE(b3 != b3a);
+    ASSERT_TRUE(b3 > b3a);
+    ASSERT_TRUE(!(b3a > b3));
+
+    bigint<3>::div_qr(quotient, remainder, b3, b2);
+    ASSERT_TRUE(quotient.num_bits() < GMP_NUMB_BITS);
+    ASSERT_TRUE(quotient.as_ulong() == b1.as_ulong());
+    ASSERT_TRUE(remainder.num_bits() < GMP_NUMB_BITS);
+    ASSERT_TRUE(remainder.as_ulong() == 42);
+
+    b3a.clear();
+    ASSERT_TRUE(b3a.is_zero());
+    ASSERT_TRUE(b3a.num_bits() == 0);
+    ASSERT_TRUE(!(b3.is_zero()));
+
+    bigint<4> bx = bigint<4>().randomize();
+    bigint<4> by = bigint<4>().randomize();
+    ASSERT_TRUE(!(bx == by));
+
+    // TODO: test serialization
+}
+
+namespace {
+    template<typename FieldT>
+    void test_field()
+    {
+        bigint<1> rand1 = bigint<1>("76749407");
+        bigint<1> rand2 = bigint<1>("44410867");
+        bigint<1> randsum = bigint<1>("121160274");
+
+        FieldT zero = FieldT::zero();
+        FieldT one = FieldT::one();
+        FieldT a = FieldT::random_element();
+        FieldT a_ser;
+        a_ser = reserialize<FieldT>(a);
+        ASSERT_TRUE(a_ser == a);
+
+        FieldT b = FieldT::random_element();
+        FieldT c = FieldT::random_element();
+        FieldT d = FieldT::random_element();
+
+        ASSERT_TRUE(a != zero);
+        ASSERT_TRUE(a != one);
+
+        ASSERT_TRUE(a * a == a.squared());
+        ASSERT_TRUE((a + b).squared() == a.squared() + a*b + b*a + b.squared());
+        ASSERT_TRUE((a + b)*(c + d) == a*c + a*d + b*c + b*d);
+        ASSERT_TRUE(a - b == a + (-b));
+        ASSERT_TRUE(a - b == (-b) + a);
+
+        ASSERT_TRUE((a ^ rand1) * (a ^ rand2) == (a^randsum));
+
+        ASSERT_TRUE(a * a.inverse() == one);
+        ASSERT_TRUE((a + b) * c.inverse() == a * c.inverse() + (b.inverse() * c).inverse());
+
+    }
+
+    template<typename FieldT>
+    void test_sqrt()
+    {
+        for (size_t i = 0; i < 100; ++i)
+        {
+            FieldT a = FieldT::random_element();
+            FieldT asq = a.squared();
+            ASSERT_TRUE(asq.sqrt() == a || asq.sqrt() == -a);
```

Comment:
This test might fail if ``sqrt()`` were nondeterministic in its choice of square-root, which AFAIK would satisfy its specification.